From: Thomas Schwinge <thomas@schwinge.name>
Subject: [PATCH] libc_once

libc_once

Needed for HEAD and glibc-2_7-branch.

Patch originally by Thomas Schwinge.  Later rewritten by Debian's Aurelien
Jarno because of upstream glibc code changes.

Reported at <http://savannah.gnu.org/bugs/?18217> and
<http://sourceware.org/bugzilla/show_bug.cgi?id=3748>, where it was rejected by
Ulrich Drepper.  Roland promised to have a look.


r2630 of debian/patches/hurd-i386/submitted-libc_once.diff.

2006-12-18  Thomas Schwinge  <tschwinge@gnu.org>

	* bits/libc-lock.h (__libc_once_else): New definiton.
	* sysdeps/mach/bits/libc-lock.h: Likewise.
	* sysdeps/mach/hurd/bits/libc-lock.h: Likewise.
	* nptl/sysdeps/pthread/bits/libc-lock.h: Likewise.
	* sysdeps/posix/getaddrinfo.c (getaddrinfo): Use __libc_once_else and a
	new local function instead of using implementational details.

Date: 2007-10-28 19:40:34 +0000 (Sun, 28 Oct 2007)
New Revision: 2630

2007-10-28  Aurelien Jarno  <aurelien@aurel32.net>

	* bits/libc-lock.h (__libc_once_get): New definiton.
	* linuxthreads/sysdeps/pthread/bits/libc-lock.h: Likewise
	* nptl/sysdeps/pthread/bits/libc-lock.h: Likewise.
	* sysdeps/mach/bits/libc-lock.h: Likewise.
	* sysdeps/mach/hurd/bits/libc-lock.h: Likewise.
	* sysdeps/posix/getaddrinfo.c (getaddrinfo): Use __libc_once_get
	  instead of using implementational details.

glibc-2.8/debian/patches/hurd-i386/submitted-libc_once.diff 3151

---
 bits/libc-lock.h                      |    3 +++
 nptl/sysdeps/pthread/bits/libc-lock.h |    3 +++
 sysdeps/mach/hurd/bits/libc-lock.h    |    4 ++++
 sysdeps/posix/getaddrinfo.c           |    2 +-
 4 files changed, 11 insertions(+), 1 deletion(-)

--- a/bits/libc-lock.h
+++ b/bits/libc-lock.h
@@ -98,6 +98,9 @@
     }									      \
   } while (0)
 
+/* Get once control variable.  */
+#define __libc_once_get(ONCE_CONTROL) \
+  ((ONCE_CONTROL) == 1)
 
 /* Start a critical region with a cleanup function */
 #define __libc_cleanup_region_start(DOIT, FCT, ARG)			    \
--- a/nptl/sysdeps/pthread/bits/libc-lock.h
+++ b/nptl/sysdeps/pthread/bits/libc-lock.h
@@ -451,6 +451,9 @@
     }									      \
   } while (0)
 
+/* Get once control variable.  */
+#define __libc_once_get(ONCE_CONTROL) \
+  ((ONCE_CONTROL) != PTHREAD_ONCE_INIT)
 
 /* Note that for I/O cleanup handling we are using the old-style
    cancel handling.  It does not have to be integrated with C++ snce
--- a/sysdeps/mach/hurd/bits/libc-lock.h
+++ b/sysdeps/mach/hurd/bits/libc-lock.h
@@ -191,6 +191,10 @@
     __libc_lock_unlock (ONCE_CONTROL.lock);				      \
   } while (0)
 
+/* Get once control variable.  */
+#define __libc_once_get(ONCE_CONTROL) \
+  (!ONCE_CONTROL.done)
+
 #ifdef _LIBC
 /* We need portable names for some functions.  E.g., when they are
    used as argument to __libc_cleanup_region_start.  */
--- a/sysdeps/posix/getaddrinfo.c
+++ b/sysdeps/posix/getaddrinfo.c
@@ -2338,7 +2338,7 @@
 	  __libc_lock_define_initialized (static, lock);
 
 	  __libc_lock_lock (lock);
-	  if (old_once && gaiconf_reload_flag)
+	  if (__libc_once_get(old_once) && gaiconf_reload_flag)
 	    gaiconf_reload ();
 	  qsort_r (order, nresults, sizeof (order[0]), rfc3484_sort, &src);
 	  __libc_lock_unlock (lock);
